src/model/model.cpp

```diff
@@ -0,0 +1,83 @@
+#include <stack>
+#include <unordered_set>
+
+#include <deepworks/model.hpp>
+
+#include "expression/call_impl.hpp"
+#include "expression/placeholder_impl.hpp"
+
+#include "model/model_impl.hpp"
+#include "model/graphbuilder.hpp"
+
+#include "util/assert.hpp"
+
+using namespace deepworks::graph;
+
+deepworks::Model::Model(deepworks::Placeholder in, deepworks::Placeholder out)
+    : deepworks::Model(deepworks::Placeholders{in}, deepworks::Placeholders{out}) {
+}
+
+deepworks::Model::Model(deepworks::Placeholders ins, deepworks::Placeholders outs)
+    : m_impl(new Impl(std::move(ins), std::move(outs))) {
+}
+
+const deepworks::Placeholders& deepworks::Model::inputs() const {
+    return m_impl->m_inputs;
+}
+
+const deepworks::Placeholders& deepworks::Model::outputs() const {
+    return m_impl->m_outputs;
+}
+
+const deepworks::Layers& deepworks::Model::layers() const {
+    return m_impl->m_layers;
+}
+
+deepworks::Layers& deepworks::Model::layers() {
+    return m_impl->m_layers;
+}
+
+deepworks::Layer deepworks::Model::getLayer(const std::string& name) {
+    auto it = m_impl->m_layers_map.find(name);
+    DeepWorks_Assert(it != m_impl->m_layers_map.end() && "Layer with that name not found");
+    return it->second;
+}
+
+deepworks::Model::Impl::Impl(deepworks::Placeholders ins,
+                             deepworks::Placeholders outs)
+    : m_tg(m_g), m_inputs(std::move(ins)), m_outputs(std::move(outs)) {
+    deepworks::GraphBuilder builder(m_g);
+    builder.build(m_inputs, m_outputs);
```

Comment:
What does this builder do? Methods do not modify the input, but they also do not return anything.

Reply:
The `builder` change input graph, because capture it by reference